ingestors/nuget.go

```diff
@@ -0,0 +1,132 @@
+package ingestors
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/librariesio/depper/data"
+)
+
+const nugetSchedule = "*/5 * * * *"
+const nugetIndexUrl = "https://api.nuget.org/v3/catalog0/index.json"
+
+type nugetIndex struct {
+	IndexId string `json:"@id"`
+	Pages   []struct {
+		Url             string `json:"@id"`
+		CommitTimeStamp string `json:"commitTimeStamp"`
+		CommitTime      time.Time
+	} `json:"items"`
+}
+
+type nugetPage struct {
+	PageId   string `json:"@id"`
+	Packages []struct {
+		Url             string `json:"@id"`
+		CommitTimeStamp string `json:"commitTimeStamp"`
+		CommitTime      time.Time
+		Name            string `json:"nuget:id"`
+		Version         string `json:"nuget:version"`
+	} `json:"items"`
+}
+
+type Nuget struct {
+	LatestRun time.Time
+}
+
+func NewNuget() *Nuget {
+	return &Nuget{}
+}
+
+func (ingestor *Nuget) Schedule() string {
+	return nugetSchedule
+}
+
+func (ingestor *Nuget) Ingest() []data.PackageVersion {
+	// Until we save LatestRun state, go back to a 4x multiple of the frequency: 20 min.
+	if ingestor.LatestRun.IsZero() {
+		ingestor.LatestRun = time.Now().Add(-20 * time.Minute)
+	}
+	packages := ingestor.ingestURL(nugetIndexUrl)
+	ingestor.LatestRun = time.Now()
+	return packages
+}
+
+func (ingestor *Nuget) ingestURL(url string) []data.PackageVersion {
+	var results []data.PackageVersion
+
+	results, err := ingestor.getIndex(url)
+	if err != nil {
+		log.WithFields(log.Fields{"ingestor": "nuget", "error": err}).Error()
+	}
+
+	return results
+}
+
+func (ingestor *Nuget) getIndex(url string) ([]data.PackageVersion, error) {
+	var results []data.PackageVersion
+
+	response, err := http.Get(url)
+	if err != nil {
+		return results, err
+	}
+	defer response.Body.Close()
+
+	var index nugetIndex
+	body, _ := ioutil.ReadAll(response.Body)
+	err = json.Unmarshal(body, &index)
+	if err != nil {
+		return results, err
+	}
+
+	for _, page := range index.Pages {
+		page.CommitTime, _ = time.Parse(time.RFC3339, page.CommitTimeStamp)
+		if page.CommitTime.After(ingestor.LatestRun) {
```

Comment:
Do we know if these times are accurate (eg, if something is published at time t, it is immediately available at time t? or is there some lag?)

Reply:
Looks like there is ~few minutes lag on the index updates, so the website will actually be more up-to-date than the index (i.e. our updates should never see missing downloads).

Also, FWIW< looks like the "commitTimestamp" from this feed is slightly later than the actual push/commit timestamps. For example, YandexMusicResolver 2.2.1:

* `2021-01-28T19:10:27Z`: timestamp from [github commit](https://github.com/SKProCH/YandexMusicResolver/commit/f410e08aebb91b1d66057ea03952de81c5a53110)
* `2021-01-28T19:11:38.7670000`: timestamp from [Nuget website version](https://www.nuget.org/packages/YandexMusicResolver/)
* `2021-01-28T19:14:05.0346383Z`: timestamp from [JSON feed](https://api.nuget.org/v3/catalog0/page11822.json)